File: lib/thick/server_handler.rb
module Thick

  class ServerHandler < Thick::Java::SimpleChannelUpstreamHandler

    def initialize(options)
      super()
      @options = options
    end

    def channelConnected(context, event)
    end

    def messageReceived(context, event)
      request = event.message
      case request
        when Thick::Java::HttpRequest
          on_request(context, request)
        when Thick::Java::HttpChunk
          on_chunk(context, request)
      end
    end

    def serve_file(context, file, from = 0, to = nil)
      raf = Thick::Java::RandomAccessFile.new(file, "r")
      to ||= File.size(file)
      region = Thick::Java::DefaultFileRegion.new(raf.channel, from, to)
      context.channel.write(region)
    end

    def on_request(context, request)

      # Serve file if exists in public directory
      file = File.join(@options[:directory], 'public', request.uri)
      return serve_file(context, file).addListener(Thick::Java::ChannelFutureListener::CLOSE) if !request.uri.index('..') && File.exists?(file) && !File.directory?(file)

      query_string = request.uri.split('?', 2)
      @env = {
          # Rack specific
          'rack.version' => [1,1],
          'rack.url_scheme' => 'http', # ToDo: support https?
          'rack.input' => Buffer.new,
          'rack.errors' => $stdout, # ToDo: Sure about that?
          'rack.multithread' => true,
          'rack.multiprocess' => false,
          'rack.run_once' => false,
          # HTTP specific
          'REQUEST_METHOD' => request.get_method.name,
          'SCRIPT_NAME' => '',
          'PATH_INFO' => query_string[0],
          'QUERY_STRING' => if query_string[1] && query_string[1] != '' then "?#{query_string[1]}" else '' end,
          'SERVER_NAME' => 'localhost', # ToDo: Be more precise!
          'SERVER_PORT' => '8080', # ToDo: Be more precise!
          # Thick specific
          'thick.async' => AsyncResponse.new(context),
          'thick.options' => @options
      }

      # Get content length if available
      length = Thick::Java::HttpHeaders.get_content_length(request)
      @env['CONTENT_LENGTH'] = length if length > 0

      # Extract headers and normalize it's names
      request.header_names.each do |name|
        rack_name = name.gsub('-', '_').upcase
        rack_name = "HTTP_#{rack_name}" unless %w(CONTENT_TYPE CONTENT_LENGTH).include?(rack_name)
        @env[rack_name] = request.get_header(name)
      end

      # Write request content into prepared IO
      @env['rack.input'].write(request.content)

      # No chunks expected, handle request
      handle_request(context) unless request.chunked?
    end

    def on_chunk(context, chunk)
      # Write request content into prepared IO
      @env['rack.input'].write(chunk.content)

      # No more chunks expected, handle request
      handle_request(context) if chunk.last?
    end

    def handle_request(context)
      response = @options[:application].call(@env)

      # Let the application make the response as it wants
      if @env['thick.async'].custom?
        @env['thick.async'].ready!
        return
      end

      # Unpack response
      status, headers, content = response

      # Create response and set status as requested by application
      @response = Thick::Java::DefaultHttpResponse.new(Thick::Java::HttpVersion::HTTP_1_1, Thick::Java::HttpResponseStatus.value_of(status.to_i))

      # Set headers as requested by application
      headers.each do |name, value|
        Thick::Java::HttpHeaders.set_header(@response, name, value)
      end

      if @env['thick.async'].async?
        if length = @env['thick.async'].content_length
          Thick::Java::HttpHeaders.set_content_length(@response, length)
        else
          @response.remove_header('Content-Length')
        end
        @response.chunked = true if @env['thick.async'].chunked?
      end

      context.channel.write(@response)

      # Set content as requested by application
      case
        when content.respond_to?(:to_path)
          serve_file(context, content.to_path)
        else
          content.each do |chunk|
            context.channel.write(Thick::Java::ChannelBuffers.copiedBuffer(chunk, Thick::Java::CharsetUtil::UTF_8))
          end
      end

      # When content is closable, close it
      content.close if content.respond_to?(:close)

      # Trigger async handler and marker if async response requested, close connection otherwise
      if @env['thick.async'].async?
        @env['thick.async'].ready!
      else
        @env['thick.async'].close
      end
    end

    def channelClosed(context, event)
      @env['thick.async'].closed!
    end

    def exceptionCaught(context, e)
      puts e.message if e.respond_to?(:message)
      puts e.backtrace if e.respond_to?(:backtrace)
    end

  end

end

File: java/src/main/java/cz/marekjelen/thick/ServerHandler.java
package cz.marekjelen.thick;

import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.SimpleChannelInboundHandler;
import io.netty.handler.codec.http.*;
import io.netty.handler.codec.http.websocketx.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.nio.charset.Charset;
import java.util.HashMap;
import java.util.Map;

public class ServerHandler extends SimpleChannelInboundHandler<Object> {

    private ServerEnvironment serverEnvironment;
    private ByteBuf inputBuffer;
    private Map<String, Object> env;
    private WebSocketEnvironment webSocketEnvironment;
    private ServerResponse response;
    private Logger logger = LoggerFactory.getLogger(ServerHandler.class);

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        super.channelActive(ctx);
        this.logger.debug("New connection");
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        super.channelInactive(ctx);
        this.logger.debug("Closed connection");
    }

    @Override
    public void channelRead0(ChannelHandlerContext context, Object message) throws Exception {
        this.logger.info(message.getClass().getName());
        if(message instanceof HttpRequest){
            onRequest(context, (HttpRequest) message);
        }else if(message instanceof HttpContent){
            onChunk(context, (HttpContent) message);
        }else if (message instanceof WebSocketFrame){
            onWebSocket(context, (WebSocketFrame) message);
        }
    }

    private void onRequest(ChannelHandlerContext context, HttpRequest request) throws Exception {

        response = new ServerResponse(context);
        inputBuffer = Unpooled.buffer();

        if(request.headers().contains("Upgrade") && request.headers().get("Upgrade").equals("websocket")){
            this.logger.info("Upgrading");
            env = buildEnvironment(request);

            env.put("PATH_INFO", "/thick/websockets"); // ToDo: path selection
            env.put("QUERY_STRING", "id=xyz"); // ToDo: id generation?
            env.put("rack.input", inputBuffer);

            DefaultFullHttpRequest req = new DefaultFullHttpRequest(request.getProtocolVersion(), request.getMethod(), request.getUri());
            req.headers().set(request.headers());

            webSocketEnvironment = new WebSocketEnvironment(context, req);
            env.put("thick.websocket", webSocketEnvironment);
            env.put("thick.response_bypass", true);

            env.put("REQUEST_METHOD", "PUT");

            serverEnvironment.getApplication().call(env);

            return;
        }

        File staticFile = new File("public", request.getUri());

        if (staticFile.isFile()) {
            response.sendFile(staticFile);
            return;
        }

        env = buildEnvironment(request);

        env.put("thick.response", response);

        String[] queryString = request.getUri().split("\\?", 2);

        env.put("rack.version", new int[] {1,1});
        env.put("rack.url_scheme", "http"); // ToDo: add support for HTTPs
        env.put("rack.errors", System.err); // ToDo: where to write errors?
        env.put("rack.multithread", true);
        env.put("rack.multiprocess", false);
        env.put("rack.run_once", false);

        env.put("REQUEST_METHOD", request.getMethod().name());
        env.put("SCRIPT_NAME", "");
        env.put("PATH_INFO", queryString[0]);
        env.put("QUERY_STRING", queryString.length == 1 ? "" : queryString[1]);

        env.put("SERVER_NAME", "localhost"); // ToDo: Be more precise!
        env.put("SERVER_PORT", "8080"); // ToDo: Be more precise!

        if(request.headers().contains(HttpHeaders.Names.CONTENT_LENGTH)){
            long content_length = HttpHeaders.getContentLength(request);
            if (content_length > 0) {
                env.put("CONTENT_LENGTH", content_length);
            }
        }

        this.logger.debug("Request processed, waiting for request body");

    }

    private Map<String, Object> buildEnvironment(HttpRequest request) {
        Map<String, Object> env = new HashMap<String, Object>();

        env.put("rack.version", new int[]{1, 1});
        env.put("rack.url_scheme", "http"); // ToDo: add support for HTTPs
        env.put("rack.errors", System.err); // ToDo: where to write errors?
        env.put("rack.multithread", true);
        env.put("rack.multiprocess", false);
        env.put("rack.run_once", false);

        HijackIO hijack = new HijackIO(null);

        env.put("rack.hijack?", true);
        env.put("rack.hijack", hijack);
        env.put("rack.hijack_io", hijack);

        env.put("SCRIPT_NAME", "");

        for(String headerName : request.headers().names()){
            String rackName = headerName.replaceAll("\\-", "_").toUpperCase();
            if (!rackName.equals("CONTENT_TYPE") && !rackName.equals("CONTENT_LENGTH")) {
                rackName = "HTTP_" + rackName;
            }
            env.put(rackName, request.headers().get(headerName));
        }

        env.put("thick.response", response);

        return env;
    }

    private void onChunk(ChannelHandlerContext context, HttpContent chunk) {
        this.logger.debug("Request body chunk processed");
        inputBuffer.writeBytes(chunk.content());
        if(chunk instanceof LastHttpContent){
            if(this.webSocketEnvironment == null) handleRequest();
        }
    }

    private void handleRequest() {
        this.logger.debug("Request finished, call Ruby application");
        env.put("rack.input", inputBuffer);
        serverEnvironment.getApplication().call(env);
    }

    private void onWebSocket(ChannelHandlerContext context, WebSocketFrame frame) {
        if (frame instanceof CloseWebSocketFrame) {
            frame.retain();
            webSocketEnvironment.closed((CloseWebSocketFrame) frame);
            return;
        }

        if (frame instanceof PingWebSocketFrame) {
            context.channel().write(new PongWebSocketFrame(frame.content()));
            return;
        }

        if (frame instanceof ContinuationWebSocketFrame){
            inputBuffer.writeBytes(frame.content());
        }else{
            inputBuffer = Unpooled.buffer();
        }

        if (frame instanceof TextWebSocketFrame){
            inputBuffer.writeBytes(frame.content());
        }

        if(frame instanceof BinaryWebSocketFrame){
            inputBuffer.writeBytes(frame.content());
        }

        if(frame.isFinalFragment()){
            webSocketEnvironment.getHandler().on_data(inputBuffer.toString(Charset.defaultCharset()));
        }
    }

    public void setEnvironment(ServerEnvironment env) {
        serverEnvironment = env;
    }
}
